### ISSUE TITLE
MPS training-loop val mAP differs ~2.6× from independent yolo val on the same best.pt — CUDA shows no discrepancy

### ISSUE THREAD

**Issue by sipie8411** [ultralytics/ultralytics#24399]
### Search before asking

- [x] I have searched the Ultralytics YOLO [issues](https://github.com/ultralytics/ultralytics/issues) and found no similar bug report.

### Ultralytics YOLO Component

Train, Val

### Bug

When training a YOLOv8 detection model on Apple Silicon with `device=mps`, the val mAP50 logged into `results.csv` during training differs sharply from re-running `yolo val` on the saved `best.pt` against the **same** val split. The same arguments on CUDA show no such discrepancy.

**Numbers from three runs on the same dataset (15-class CCTV defect detection, 771 train images, project-level GroupKFold split):**

| Source | Device | best epoch | training-time val mAP50 | independent val mAP50 |
|---|---|---|---|---|
| MPS run (M4 Pro, batch=16, patience=100) | MPS | **51** | 0.297 | **0.780** |
| CUDA run #1 (RTX 5060, batch=16, patience=80) | CUDA | 17 | 0.322 | 0.317 |
| CUDA run #2 (RTX 5060, batch=8, patience=100) | CUDA | 20 | 0.307 | 0.285 |

Ratio between training-time val and independent val: **2.63× on MPS, 0.98× and 0.93× on CUDA** (two independent CUDA runs).

**Independent test set (project-aware GroupKFold, no project overlap with train/val):**

| Source | Device | independent test mAP50 |
|---|---|---|
| MPS `best.pt` evaluated on `device=mps` | MPS | **0.666** |
| CUDA #1 `best.pt` evaluated on `device=mps` | (eval on MPS) | 0.147 |
| CUDA #2 `best.pt` evaluated on `device=mps` | (eval on MPS) | 0.148 |

The same evaluation environment (same machine, same `yolo val` command, same test split, same `device=mps`) produces 0.15 for two independent CUDA-trained checkpoints and 0.67 for the MPS-trained checkpoint. The differing factor is the training device, not the evaluation device.

**Controls already eliminated**

1. **Data leakage** — train/val/test split is project-level GroupKFold (8 mutually exclusive projects: train={XHL,LSL,JXL,YSL}, val={ZSJ,XYG}, test={DCH,XSJ}), zero file overlap, verified by file listings on disk.
2. **Validator code path** — both CUDA `best.pt` checkpoints match their training-time numbers when re-evaluated independently (0.322 → 0.317 and 0.307 → 0.285). The standalone `yolo val` path is internally consistent.
3. **Numerical precision at eval time** — same MPS `best.pt` evaluated on `device=mps` (FP16), `device=mps` (FP32) and `device=cpu` produces identical mAP50 (delta < 0.001). Not an FP16 artifact at eval.
4. **EMA selection / weight save** — `strip_optimizer`-saved EMA at the recorded best epoch and `weights/last.pt` reproduce the same pattern.
5. **Dataset cache** — cache files predate the run; verified clean.
6. **`patience` as a confound** — CUDA #1 used `patience=80`, CUDA #2 used `patience=100`. Both still landed at best epoch 17–20 and independent test mAP50 ≈ 0.15. The MPS run used `patience=100` (same as CUDA #2) but landed at best epoch 51 and test mAP50 = 0.67. Patience is not the differentiating variable.
7. **`batch` as a confound** — CUDA #1 batch=16, CUDA #2 batch=8, MPS batch=16. Two CUDA runs match each other (0.147 / 0.148) across different batch sizes; the MPS run differs (0.666) at the same batch size as CUDA #1.

The remaining systematic difference between runs that produce 0.15 vs 0.67 on the independent test set is the training device. The issue appears specific to the **training loop on MPS** — not to the standalone validator and not to the saved weights themselves (the weights load and run reproducibly across CPU / MPS / CUDA at eval time).

### Environment

```
# MPS box (where the discrepancy is observed)
Ultralytics 8.4.46 🚀 Python-3.12.13 torch-2.11.0 CPU (Apple M4 Pro)
Setup complete ✅ (12 CPUs, 48.0 GB RAM, 134.1/460.4 GB disk)

OS                     macOS-26.4.1-arm64-arm-64bit
Environment            Darwin
Python                 3.12.13
Install                git
Path                   /opt/homebrew/lib/python3.12/site-packages/ultralytics
RAM                    48.00 GB
CPU                    Apple M4 Pro
GPU                    None  (note: yolo checks does not list MPS; torch.backends.mps.is_available() == True)
CUDA                   None

torch                  2.11.0
ultralytics            8.4.46  (latest on PyPI as of 2026-05-03)
```

```
# CUDA reference box #2
Ultralytics 8.4.37  Python-3.9.13 torch-2.8.0+cu128 CUDA:0 (NVIDIA GeForce RTX 5060 Laptop GPU, 8151 MiB)
OS                     Windows-10-10.0.26200-SP0
Python                 3.9.13
torch                  2.8.0+cu128
ultralytics            8.4.37
```

(CUDA reference box #1 was a different RTX 5060 desktop, also Windows + ultralytics 8.4.46. Same dataset, same args except `device` and `patience`.)

### Minimal Reproducible Example

```bash
# 1. MPS training — logs val mAP50 ≈ 0.30 in results.csv at best epoch
yolo detect train \
  model=yolov8s.pt data=data.yaml \
  epochs=200 patience=100 batch=16 imgsz=640 \
  device=mps seed=0 deterministic=True amp=True \
  optimizer=auto lr0=0.01 warmup_epochs=3.0 \
  close_mosaic=10 mosaic=1.0 augment=False half=False

# 2. Independent eval on the same val split — returns mAP50 ≈ 0.78 (2.6× of training-time number)
yolo val \
  model=runs/detect/train/weights/best.pt \
  data=data.yaml split=val device=mps imgsz=640 batch=16

# 3. For comparison, identical args on CUDA produce training-time vs independent val within ±0.04
```

The dataset cannot be shared (proprietary CCTV footage), but the split metadata, class definitions, full `args.yaml`, full `results.csv` (151 epochs for the MPS run, 159 epochs for CUDA #2), and all three `best.pt` files (≈22 MB each) can be provided on request.

### Additional

Not assuming a root cause. Reporting because the 2.63× training-time-vs-independent val gap, if reproducible, would silently mislead anyone using `device=mps` for early stopping or model selection on Apple Silicon — early-stopping decisions on the MPS-side `results.csv` numbers may stop training prematurely or, conversely, let it run far past the actual fitness peak.

A second, independent CUDA reproducer (RTX 5060 Laptop, batch=8, patience=100) was added after the initial single-CUDA-run report, specifically to rule out `patience` and `batch` as the differentiating factor. Both CUDA runs converge to the same independent test mAP50 (0.147 and 0.148). The MPS run on identical args (except device) produces 0.666 on the same test split.

Questions for maintainers:

- Does training-loop validation on MPS use a different config or code path than the standalone validator (e.g. different model/EMA mode, different conf/iou thresholds, different normalization, different `model.eval()` state)?
- Any known open issues for MPS training-time validation diverging from `yolo val`?
- If this is expected on MPS, could it be noted in the `device=mps` documentation?

Happy to provide any of the listed artifacts or run diagnostic commands.

### Are you willing to submit a PR?

- [x] #24400

**Comment by UltralyticsAssistant** [comment]
👋 Hello @sipie8411, thank you for the detailed bug report and for testing this so thoroughly 🚀 This is an automated response to help move things forward quickly, and an Ultralytics engineer will also assist soon.

We recommend a visit to the [Docs](https://docs.ultralytics.com/) for new users where you can find many [Python](https://docs.ultralytics.com/usage/python/) and [CLI](https://docs.ultralytics.com/usage/cli/) usage examples and where many of the most common questions may already be answered.

Since this appears to be a 🐛 Bug Report, thanks for already including a strong reproducibility outline. To help the team investigate further, please also provide a fully self-contained [minimum reproducible example](https://docs.ultralytics.com/help/minimum-reproducible-example/) if you can, including the exact command history, relevant `args.yaml`, `results.csv`, and any logs showing the train-time validation metrics versus the standalone `yolo val` metrics on the same checkpoint 📋

If this is a custom training ❓ Question, please provide as much information as possible, including dataset image examples and training logs, and verify you are following our [Tips for Best Training Results](https://docs.ultralytics.com/guides/model-training-tips/).

Join the Ultralytics community where it suits you best. For real-time chat, head to [Discord](https://discord.com/invite/ultralytics) 🎧. Prefer in-depth discussions? Check out [Discourse](https://community.ultralytics.com/). Or dive into threads on our [Subreddit](https://www.reddit.com/r/Ultralytics/) to share knowledge with the community.

## Upgrade

Please also verify the issue on the latest `ultralytics` package including all [requirements](https://github.com/ultralytics/ultralytics/blob/main/pyproject.toml) in a [**Python>=3.8**](https://www.python.org/) environment with [**PyTorch>=1.8**](https://pytorch.org/get-started/locally/), as this helps confirm the behavior is still present in the most recent release:

```bash
pip install -U ultralytics
```

## Environments

YOLO may be run in any of the following up-to-date verified environments (with all dependencies including [CUDA](https://developer.nvidia.com/cuda)/[CUDNN](https://developer.nvidia.com/cudnn), [Python](https://www.python.org/) and [PyTorch](https://pytorch.org/) preinstalled):

- **Notebooks** with free GPU: <a href="https://console.paperspace.com/github/ultralytics/ultralytics"><img src="https://assets.paperspace.io/img/gradient-badge.svg" alt="Run on Gradient"/></a> <a href="https://colab.research.google.com/github/ultralytics/ultralytics/blob/main/examples/tutorial.ipynb"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"></a> <a href="https://www.kaggle.com/models/ultralytics/yolo26"><img src="https://kaggle.com/static/images/open-in-kaggle.svg" alt="Open In Kaggle"></a>
- **Google Cloud** Deep Learning VM. See [GCP Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/google_cloud_quickstart_tutorial/)
- **Amazon** Deep Learning AMI. See [AWS Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/aws_quickstart_tutorial/)
- **Docker Image**. See [Docker Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/docker_image_quickstart_tutorial/) <a href="https://hub.docker.com/r/ultralytics/ultralytics"><img src="https://img.shields.io/docker/pulls/ultralytics/ultralytics?logo=docker" alt="Docker Pulls"></a>

## Status

<a href="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule"><img src="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml/badge.svg" alt="Ultralytics CI"></a>

If this badge is green, all [Ultralytics CI](https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule) tests are currently passing. CI tests verify correct operation of all YOLO [Modes](https://docs.ultralytics.com/modes/) and [Tasks](https://docs.ultralytics.com/tasks/) on macOS, Windows, and Ubuntu every 24 hours and on every commit.

For this report specifically, the most helpful next additions would be:
- The full `args.yaml` from the MPS run and one CUDA comparison run
- The relevant `results.csv` files
- The console output from both the train run and the standalone `yolo val`
- Confirmation whether the same behavior appears after upgrading with `pip install -U ultralytics`
- If possible, a small sanitized reproduction or artifact bundle that follows the [minimum reproducible example](https://docs.ultralytics.com/help/minimum-reproducible-example/) guide 🔍

**Comment by glenn-jocher** [comment]
Thanks for the detailed repro — this is not expected. Training-loop val and standalone `yolo val` use the same validator, but not the exact same path: during training [`BaseValidator.__call__()`](https://docs.ultralytics.com/reference/engine/validator/) validates `trainer.ema.ema or trainer.model` and sets `self.args.half = self.device.type != "cpu" and trainer.amp`, while standalone val goes through `AutoBackend`; that makes MPS + `amp=True` the first thing to isolate here, especially since [`check_amp()`](https://docs.ultralytics.com/reference/utils/checks/) treats AMP as CUDA-only. Please retry once with `amp=False`, then attach `yolo checks`, `ultralytics.utils.checks.collect_system_info()` from the imported package, and if possible a minimal reproducible sample with `best.pt`, `args.yaml`, and `results.csv` so we can reproduce this locally.

**Comment by Y-T-G** [comment]
Hello. Thanks for the detailed report.

Did you try CPU training? Does it match CUDA or MPS results? This looks like AMP issue.

Can you provide the training plots from each training showing the mAP vs. epoch?

**Comment by sipie8411** [comment]
`Update:` ran CPU training on Mac mini (M4 Pro) with same dataset, seed=0, patience=80; only `amp=False` (CPU can't use AMP) and `workers=4` (vs CUDA 8) differ.

Training stopped at epoch 101 via patience. **Best epoch 21, mAP50 = 0.307**.

Top 5 best epochs from the CPU run:

| epoch | mAP50 | mAP50-95 |
|---|---|---|
| 21 | 0.307 | (best) |
| 7  | 0.267 | |
| 27 | 0.261 | |
| 23 | 0.258 | |
| 44 | 0.245 | |

Three-backend matrix now:

| Backend | Setup | Best mAP50 |
|---|---|---|
| CUDA (office baseline) | amp=True, workers=8 | 0.150 |
| CPU (mini, this test) | amp=False, workers=4 | **0.307** |
| MPS (mini, original anomaly) | amp=True, workers=8 | **0.666** |

Three backends, three different results — MPS remains the most extreme outlier.

CPU diverges from CUDA by ~2x. Possible explanations: AMP off vs on, fewer workers (different shuffle ordering), and CPU/CUDA deterministic-mode implementations may differ at the kernel level. On a small 494-image training set, ~2x variance from these config differences is plausible.

What's *not* plausible: MPS at 0.666 — that's 4x+ over CUDA and 2x+ over CPU. The MPS divergence is materially larger than what amp/workers/deterministic differences would explain.

Happy to share `args.yaml`, full `results.csv`, and training curve plots (results.png) — let me know what's most useful.

<img width="2400" height="1200" alt="Image" src="https://github.com/user-attachments/assets/caa95b3a-d166-4609-aaf7-c4d37e6a01df" />

**Comment by glenn-jocher** [comment]
Thanks, this CPU result is useful because it points more toward a training-backend issue than a standalone validator issue; the next key A/B is rerunning the same MPS training with `amp=False` and otherwise identical args, since training-time val inherits `trainer.amp` in [`BaseValidator.__call__()`](https://docs.ultralytics.com/reference/engine/validator/) and [`check_amp()`](https://docs.ultralytics.com/reference/utils/checks/) is CUDA-oriented, and if you can attach the MPS `amp=False` `args.yaml`, `results.csv`, `results.png`, plus `ultralytics.utils.checks.collect_system_info()` output from that exact environment, that should give us the clearest repro.

**Comment by mulhamfetna** [comment]
I've worked on cross-platform ML training issues.

**Technical Analysis:**
mAP differs 2.6x between training-loop validation and independent inference suggests:
1. Batch normalization statistics not synced
2. Model weights loaded incorrectly on MPS
3. Different precision modes between train and inference

**Root Cause (likely):**
- MPS uses different precision handling than CUDA
- BatchNorm may compute different statistics on MPS vs CPU
- Validation during training vs after training uses different code paths

**Fix Strategy:**
1. Force eval() mode consistently
2. Check BN statistics handling for MPS
3. Verify weight loading preserves precision

Note: CUDA shows no discrepancy, so this is MPS-specific.

**Comment by mulhamfetna** [comment]
I've analyzed the issue and created a fix. The root cause appears to be related to how the model is converted to FP32 during training validation on MPS.

The fix:
1. Excludes MPS from half precision during training validation to prevent the mAP discrepancy
2. Uses `.to(torch.float32)` instead of `.float()` for more reliable FP32 conversion on MPS

PR: https://github.com/ultralytics/ultralytics/pull/24439

**Comment by mulhamfetna** [comment]
Thank you for the review. Addressing the same issues identified in related PRs. Will update shortly.

**Comment by glenn-jocher** [comment]
Thanks for the update — please keep the PR minimal and focused on the MPS validation mismatch, add a regression test covering training-loop val vs standalone `yolo val`, and avoid introducing new `default.yaml` args unless they’re strictly required.
